src/lib/profiles/software-update/SoftwareUpdateProfile.cpp

```diff
@@ -28,460 +28,575 @@
- * return: error/status
+/**
+ * Explicitly initialize the IntegritySpec object with provided values.
+ *
+ * @param[in] aType An integrity type value drawn from @ref IntegrityTypes
+ * @param[in] aValue A hash value of the appropriate length represented as a packed string of bytes
+ *
+ * @return WEAVE_NO_ERROR On success
+ * @return WEAVE_ERROR_INVALID_INTEGRITY_TYPE If the provided integrity type is not one of the values specified in @ref
+ * IntegrityTypes
  */
-WEAVE_ERROR IntegritySpec::init(uint8_t aType, uint8_t* aValue)
+WEAVE_ERROR IntegritySpec::init(uint8_t aType, uint8_t * aValue)
 {
-  uint8_t len=integrityLength(aType);
-  if (len==0) return WEAVE_ERROR_INVALID_INTEGRITY_TYPE;
-  type=aType;
-  for (int i=0; i<len; i++) value[i]=aValue[i];
-  return WEAVE_NO_ERROR;
+    uint8_t len = integrityLength(aType);
+    if (len == 0)
+        return WEAVE_ERROR_INVALID_INTEGRITY_TYPE;
+    type = aType;
+    for (int i = 0; i < len; i++)
+        value[i] = aValue[i];
+    return WEAVE_NO_ERROR;
 }
-/*
- * parameters: MessageIterator &i, an iterator ofver the message
- * being packed
- * returns: error/status
+/**
+ * Serialize the IntegritySpec into provided MessageIterator
+ * @param[in] i An iterator over the message being packed
+ *
+ * @retval WEAVE_NO_ERROR  On success.
+ * @retval WEAVE_ERROR_BUFFER_TOO_SMALL If the IntegritySpec is too large to fit in the message.
  */
-WEAVE_ERROR IntegritySpec::pack(MessageIterator &i)
+WEAVE_ERROR IntegritySpec::pack(MessageIterator & i)
 {
-  TRY(i.writeByte(type));
-  for (int j=0; j<integrityLength(type); j++) TRY(i.writeByte(value[j]));
-  return WEAVE_NO_ERROR;
+    TRY(i.writeByte(type));
+    for (int j = 0; j < integrityLength(type); j++)
+        TRY(i.writeByte(value[j]));
+    return WEAVE_NO_ERROR;
 }
-/*
- * parameters:
- * - MessageIterator &i, an iterator over the message being parsed
- * - IntegritySpec &aSpec, a pointer to an object to contain the result
- * returns: error/status
+/**
+ * Deserialize the object from the provided MessageIterator into provided IntegritySpec
+ * @param[in] i An iterator over the message being parsed.
+ * @param[in] aSpec A reference to an object to contain the result
+ *
+ * @retval WEAVE_NO_ERROR  On success.
+ * @return WEAVE_ERROR_INVALID_INTEGRITY_TYPE If the provided integrity type is not one of the values specified in @ref
+ * IntegrityTypes
+ * @retval WEAVE_ERROR_BUFFER_TOO_SMALL If the message did not contain enough bytes for the integrity type and the associated hash
  */
-WEAVE_ERROR IntegritySpec::parse(MessageIterator &i, IntegritySpec &aSpec)
+WEAVE_ERROR IntegritySpec::parse(MessageIterator & i, IntegritySpec & aSpec)
 {
-  TRY(i.readByte(&aSpec.type));
-  for (int j=0; j<integrityLength(aSpec.type); j++) TRY(i.readByte(&aSpec.value[j]));
-  return WEAVE_NO_ERROR;
+    TRY(i.readByte(&aSpec.type));
+    if (integrityLength(aSpec.type) == 0)
+        return WEAVE_ERROR_INVALID_INTEGRITY_TYPE;
```

Comment:
Note: New error condition, error is much more consistent with other similar code paths

Reply:
Assertion?